`flutter run` fails in a very ugly way if you haven't signed Android Studio EULAs

```
$ flutter run
Launching lib/main.dart on Nexus 6 in debug mode...
Running 'gradle assembleDebug'...                         
Checking the license for package Android SDK Build-Tools 24.0.1 in /usr/local/google/home/ianh/Android/Sdk/licenses
Warning: License for package Android SDK Build-Tools 24.0.1 not accepted.

FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring project ':app'.
> Failed to notify project evaluation listener.
   > You have not accepted the license agreements of the following SDK components:
     [Android SDK Build-Tools 24.0.1].
     Before building your project, you need to accept the license agreements and complete the installation of the missing components using the Android Studio SDK Manager.
     Alternatively, to learn how to transfer the license agreements from one workstation to another, go to http://d.android.com/r/studio-ui/export-licenses.html
   > Could not get unknown property 'compileDebugJavaWithJavac' for project ':app' of type org.gradle.api.Project.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.
Exception from flutter run: Exception: Gradle build failed: 1
package:flutter_tools/src/base/common.dart 29                     throwToolExit
package:flutter_tools/src/android/gradle.dart 183                 buildGradleProjectV2
package:flutter_tools/src/android/gradle.dart 133                 buildGradleProject
package:flutter_tools/src/commands/build_apk.dart 592             buildAndroidWithGradle
package:flutter_tools/src/commands/build_apk.dart 603             buildApk
package:flutter_tools/src/android/android_device.dart 291         AndroidDevice.startApp
package:flutter_tools/src/run_hot.dart 165                        HotRunner._run
package:flutter_tools/src/run_hot.dart 77                         HotRunner.run.<fn>
package:stack_trace                                               Chain.capture
package:flutter_tools/src/run_hot.dart 76                         HotRunner.run
package:flutter_tools/src/commands/run.dart 284                   RunCommand.runCommand
package:flutter_tools/src/runner/flutter_command.dart 148         FlutterCommand.verifyThenRunCommand
package:flutter_tools/src/commands/run.dart 199                   RunCommand.verifyThenRunCommand
package:flutter_tools/src/runner/flutter_command.dart 119         FlutterCommand.run
package:args/command_runner.dart 194                              CommandRunner.runCommand
package:flutter_tools/src/runner/flutter_command_runner.dart 224  FlutterCommandRunner.runCommand
package:args/command_runner.dart 109                              CommandRunner.run.<fn>
dart:async                                                        Future.Future.sync
package:args/command_runner.dart 109                              CommandRunner.run
package:flutter_tools/src/runner/flutter_command_runner.dart 151  FlutterCommandRunner.run
package:flutter_tools/executable.dart 128                         main.<fn>.<fn>
package:stack_trace                                               Chain.capture
package:flutter_tools/executable.dart 127                         main.<fn>
package:flutter_tools/src/base/context.dart 76                    AppContext._run
package:flutter_tools/src/base/context.dart 66                    AppContext.runInZone.<fn>
dart:async                                                        runZoned
package:flutter_tools/src/base/context.dart 65                    AppContext.runInZone
package:flutter_tools/executable.dart 98                          main
../../packages/flutter_tools/bin/flutter_tools.dart 8             main
===== asynchronous gap ===========================
dart:async                                                        _Completer.completeError
package:flutter_tools/src/android/gradle.dart 196                 buildGradleProjectV2
===== asynchronous gap ===========================
dart:async                                                        _asyncThenWrapperHelper
package:flutter_tools/src/android/gradle.dart                     buildGradleProjectV2
package:flutter_tools/src/android/gradle.dart 133                 buildGradleProject
===== asynchronous gap ===========================
dart:async                                                        _asyncThenWrapperHelper
package:flutter_tools/src/android/gradle.dart                     buildGradleProject
package:flutter_tools/src/commands/build_apk.dart 592             buildAndroidWithGradle
===== asynchronous gap ===========================
dart:async                                                        Future.Future.microtask
package:flutter_tools/src/commands/build_apk.dart                 buildAndroidWithGradle
package:flutter_tools/src/commands/build_apk.dart 603             buildApk
===== asynchronous gap ===========================
dart:async                                                        Future.Future.microtask
package:flutter_tools/src/commands/build_apk.dart                 buildApk
package:flutter_tools/src/android/android_device.dart 291         AndroidDevice.startApp
===== asynchronous gap ===========================
dart:async                                                        _asyncThenWrapperHelper
package:flutter_tools/src/android/android_device.dart             AndroidDevice.startApp
package:flutter_tools/src/run_hot.dart 165                        HotRunner._run
===== asynchronous gap ===========================
dart:async                                                        _asyncThenWrapperHelper
package:flutter_tools/src/run_hot.dart                            HotRunner._run
package:flutter_tools/src/run_hot.dart 77                         HotRunner.run.<fn>
package:stack_trace                                               Chain.capture
package:flutter_tools/src/run_hot.dart 76                         HotRunner.run
package:flutter_tools/src/commands/run.dart 284                   RunCommand.runCommand

```

Current output is still pretty terrible. It's now:

```
Launching lib/main.dart on Nexus 6 in debug mode...
FAILURE: Build failed with an exception.

* Where:
Build file '/Users/ianh/dev/dummy/android/build.gradle' line: 20

* What went wrong:
A problem occurred evaluating root project 'android'.
> A problem occurred configuring project ':app'.
   > Failed to notify project evaluation listener.
      > You have not accepted the license agreements of the following SDK components:
        [Android SDK Build-Tools 25.0.2, Android SDK Platform 25].
        Before building your project, you need to accept the license agreements and complete the installation of the missing components using the Android Studio SDK Manager.
        Alternatively, to learn how to transfer the license agreements from one workstation to another, go to http://d.android.com/r/studio-ui/export-licenses.html
      > Could not get unknown property 'compileDebugJavaWithJavac' for project ':app' of type org.gradle.api.Project.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.
Error running gradle: Exit code 1 from: /Applications/Android Studio.app/Contents/gradle/gradle-3.2/bin/gradle app:properties
Running 'gradle assembleDebug'...                         
Checking the license for package Android SDK Build-Tools 25.0.2 in /Users/ianh/Library/Android/sdk/licenses
Warning: License for package Android SDK Build-Tools 25.0.2 not accepted.
Checking the license for package Android SDK Platform 25 in /Users/ianh/Library/Android/sdk/licenses
Warning: License for package Android SDK Platform 25 not accepted.

FAILURE: Build failed with an exception.

* Where:
Build file '/Users/ianh/dev/dummy/android/build.gradle' line: 20

* What went wrong:
A problem occurred evaluating root project 'android'.
> A problem occurred configuring project ':app'.
   > Failed to notify project evaluation listener.
      > You have not accepted the license agreements of the following SDK components:
        [Android SDK Build-Tools 25.0.2, Android SDK Platform 25].
        Before building your project, you need to accept the license agreements and complete the installation of the missing components using the Android Studio SDK Manager.
        Alternatively, to learn how to transfer the license agreements from one workstation to another, go to http://d.android.com/r/studio-ui/export-licenses.html
      > Could not get unknown property 'compileDebugJavaWithJavac' for project ':app' of type org.gradle.api.Project.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.
Gradle build failed: 1
```

I just tested this on a squeaky clean machine (using a VM). I performed the following install steps:

1. flutter git clone
1. flutter doctor
1. Android Studio install
1. Android Studio Setup Wizard (installed SDK + tools + gradle)

After this, `flutter doctor` reported (omitting iOS section):

```
dhcp-100-105-163-111:~ user$ flutter doctor
[✓] Flutter (on Mac OS X 10.12.2 16C68, channel master)
    • Flutter at /Users/user/flutter
    • Framework revision f52cbe8189 (4 hours ago), 2017-05-08 00:06:28 -0700
    • Engine revision 1b01211ee6
    • Tools Dart version 1.23.0-dev.11.11

[✓] Android toolchain - develop for Android devices (Android SDK 25.0.3)
    • Android SDK at /Users/user/Library/Android/sdk
    • Platform android-25, build-tools 25.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version: OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] Android Studio (version 2.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Gradle version 3.2
    • Java version: OpenJDK Runtime Environment (build 1.8.0_112-release-b06)
```

However, a build still failed:

```
flutter create test01
cd test01
flutter build apk
```

=>

```
$ flutter build apk
Running 'gradle assembleRelease'...                       
Checking the license for package Android SDK Build-Tools 25.0.2 in /Users/user/Library/Android/sdk/licenses
Warning: License for package Android SDK Build-Tools 25.0.2 not accepted.

FAILURE: Build failed with an exception.

* Where:
Build file '/Users/user/tmp/test01/android/build.gradle' line: 20

* What went wrong:
A problem occurred evaluating root project 'android'.
> A problem occurred configuring project ':app'.
   > Failed to notify project evaluation listener.
      > You have not accepted the license agreements of the following SDK components:
        [Android SDK Build-Tools 25.0.2].
        Before building your project, you need to accept the license agreements and complete the installation of the missing components using the Android Studio SDK Manager.
        Alternatively, to learn how to transfer the license agreements from one workstation to another, go to http://d.android.com/r/studio-ui/export-licenses.html
      > Could not get unknown property 'compileDebugJavaWithJavac' for project ':app' of type org.gradle.api.Project.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.
Gradle build failed: 1
dhcp-100-105-163-111:test01 user$ 
```

I just did a new install into a Xubuntu virtual Machine. This is what I got today:


Initializing gradle...                                0.9s
Running 'gradlew assembleRelease'...
Checking the license for package Android SDK Build-Tools 26 in /home/ride4sun/Android/Sdk/licenses
Warning: License for package Android SDK Build-Tools 26 not accepted.

FAILURE: Build failed with an exception.

* Where:
Build file '/home/ride4sun/projects/lookatmybaby/android/build.gradle' line: 23

* What went wrong:
A problem occurred evaluating root project 'android'.
> A problem occurred configuring project ':app'.
   > You have not accepted the license agreements of the following SDK components:
     [Android SDK Build-Tools 26].
     Before building your project, you need to accept the license agreements and complete the installation of the missing components using the Android Studio SDK Manager.
     Alternatively, to learn how to transfer the license agreements from one workstation to another, go to http://d.android.com/r/studio-ui/export-licenses.html

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.
Gradle build failed: 1

--- a/packages/flutter_tools/lib/src/android/gradle.dart
+++ b/packages/flutter_tools/lib/src/android/gradle.dart
@@ -98,14 +98,39 @@ Future<GradleProject> _readGradleProject() async {
     return project;
   } catch (e) {
     if (flutterPluginVersion == FlutterPluginVersion.managed) {
-      printError('Error running Gradle:\n$e\n');
-      throwToolExit(
-        'Please review your Gradle project setup in the android/ folder.',
-      );
+      // Handle known exceptions. This will exit if handled.
+      handleKnownGradleExceptions(e);
+      
+      // Print a general Gradle error and exit.
+      printError('* Error running Gradle:\n$e\n');
+      throwToolExit('Please review your Gradle project setup in the android/ folder.');
     }
   }
   // Fall back to the default
   return new GradleProject(<String>['debug', 'profile', 'release'], <String>[], gradleAppOutDirV1);
+}
+
+void handleKnownGradleExceptions(String exceptionString) {
+  // Handle Gradle error thrown when Gradle needs to download additional
+  // Android SDK components (e.g. Platform Tools), and the license
+  // for that component has not been accepted.
+  final String matcher =
+    r'You have not accepted the license agreements of the following SDK components:'
+    r'\s*\[(.+)\]';
+  final RegExp licenseFailure = new RegExp(matcher, multiLine: true);
+  final Match licenseMatch = licenseFailure.firstMatch(exceptionString);
+  if (licenseMatch != null) {
+    final String missingLicenses = licenseMatch.group(1);
+    if (matcher.isNotEmpty) {
+      final String errorMessage =
+        '\n\n* Error running Gradle:\n'
+        'Unable to download needed Android SDK components, as the following licenses have not been accepted:\n'
+        '$missingLicenses\n\n'
+        'To resolve this, please run the following command in a Terminal:\n'
+        'flutter doctor --android-licenses';
+      throwToolExit(errorMessage);
+    }
+  }
 }
 
 String _locateProjectGradlew({ bool ensureExecutable: true }) {

--- a/packages/flutter_tools/lib/src/android/gradle.dart
+++ b/packages/flutter_tools/lib/src/android/gradle.dart
@@ -121,15 +121,13 @@ void handleKnownGradleExceptions(String exceptionString) {
   final Match licenseMatch = licenseFailure.firstMatch(exceptionString);
   if (licenseMatch != null) {
     final String missingLicenses = licenseMatch.group(1);
-    if (matcher.isNotEmpty) {
-      final String errorMessage =
-        '\n\n* Error running Gradle:\n'
-        'Unable to download needed Android SDK components, as the following licenses have not been accepted:\n'
-        '$missingLicenses\n\n'
-        'To resolve this, please run the following command in a Terminal:\n'
-        'flutter doctor --android-licenses';
-      throwToolExit(errorMessage);
-    }
+    final String errorMessage =
+      '\n\n* Error running Gradle:\n'
+      'Unable to download needed Android SDK components, as the following licenses have not been accepted:\n'
+      '$missingLicenses\n\n'
+      'To resolve this, please run the following command in a Terminal:\n'
+      'flutter doctor --android-licenses';
+    throwToolExit(errorMessage);
   }
 }
 

